File: brainfuck.c
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define STACK_SIZE 100
#define MAX_MEN 30000

typedef struct Program {
  char *data;
  size_t len;
} Program;

Program read_program(const char *filename) {
  FILE *f;
  size_t size;
  char *buf;

  f = fopen(filename, "rb");
  if (!f) {
    printf("File not found\n");
    exit(1);
  }

  fseek(f, 0, SEEK_END);
  size = ftell(f);
  fseek(f, 0, SEEK_SET);

  buf = malloc(sizeof(char) * size + 1);
  if (!(buf)) {
    printf("Error allocating a buffer for program");
    fclose(f);
    exit(1);
  }

  fread(buf, 1, size, f);
  buf[size] = '\0';
  fclose(f);
  return (Program){.data = buf, .len = size};
}

int *create_jump_table(const Program program) {
  int *table = calloc(program.len, sizeof(int));

  int stack[STACK_SIZE] = {0};
  int top = 0;

  for (int p = 0; p < program.len; p++) {
    if (*(program.data + p) == '[') {
      assert(top < STACK_SIZE);
      stack[top++] = p;
    } else if (*(program.data + p) == ']') {
      *(table + p) = stack[--top];
      *(table + stack[top]) = p;
    }
  }

  assert(top == 0);

  return table;
}

int main(int argc, char **argv) {
  if (argc < 2) {
    puts("Usage: brainfuck <program>");
    exit(1);
  }
  Program program = read_program(argv[1]);

  int pointer = 0;
  int pc = 0;
  unsigned char memory[MAX_MEN] = {0};

  int *table = create_jump_table(program);

  while (pc < program.len) {
    unsigned char *p_memory = memory + pointer;
    char inst = *(program.data + pc);

    if (inst == '<') {
      pointer--;
    } else if (inst == '>') {
      pointer++;
    } else if (inst == '+') {
      *p_memory = (*p_memory + 1) & 0xff;
    } else if (inst == '-') {
      *p_memory = (*p_memory - 1) & 0xff;
    } else if (inst == '[' && *p_memory == 0) {
      pc = *(table + pc);
    } else if (inst == ']' && *p_memory != 0) {
      pc = *(table + pc);
    } else if (inst == '.') {
      putchar(*p_memory);
    } else if (inst == ',') {
      *p_memory = getchar();
    }
    pc++;
  }
  free(program.data);
  free(table);
  return 0;
}

File: brainfuck.go
package main

import (
	"fmt"
	"log"
	"os"
	"slices"
)

func createJumpTable(program []byte) map[int]int {

	top := 0
	var stack []int
	jump := make(map[int]int)

	for ord, c := range program {
		r := string(c)
		if r == "[" {
			stack = append(stack, ord)
			top++
		} else if r == "]" {
			i := stack[len(stack)-1]
			jump[i] = ord
			jump[ord] = i
			stack = stack[:len(stack)-1]
			top--
		}
	}

	if top != 0 {
		log.Fatal("Error: [] are not balanced in this program")
	}

	return jump
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: brainfuck <filename>")
		os.Exit(0)
	}

	program, err := os.ReadFile(os.Args[1])
	if err != nil {
		log.Fatal(err)
	}

	jumpTable := createJumpTable(program)

	memory := make([]int, 30000)
	pointer := 0
	pc := 0
	inst_set := []string{"<", ">", "+", "-", "[", "]", "."}

	for pc < len(program) {

		instr := string(program[pc])
		if !slices.Contains(inst_set, instr) {
			pc++
			continue
		}

		// for debug purposes:
		//fmt.Printf("pc: %d ins: %s ptr: %d\n", pc, string(program[pc]), pointer)
		switch instr {
		case "<":
			pointer--
		case ">":
			pointer++
		case "+":
			memory[pointer] = (memory[pointer] + 1) & 0xFF
		case "-":
			memory[pointer] = (memory[pointer] - 1) & 0xFF
		case ".":
			fmt.Printf("%c", memory[pointer])
		case "[":
			if memory[pointer] == 0 {
				pc = jumpTable[pc]
			}
		case "]":
			if memory[pointer] != 0 {
				pc = jumpTable[pc]
			}
		}

		pc++
	}
}
